lib/moped/sockets/ssl.rb

```diff
@@ -45,6 +49,63 @@ def set_encoding(string)
       def setsockopt(*args)
         socket.setsockopt(*args)
       end
+
+    private
+
+      def create_context(options)
+        context = OpenSSL::SSL::SSLContext.new
+        set_cert(context, options)
+        set_key(context, options)
+        set_cert_verification(context, options) unless options[:ssl_verify] == false
+        context
+      end
+
+      def set_cert(context, options)
+        if options[:ssl_cert]
+          context.cert = OpenSSL::X509::Certificate.new(File.open(options[:ssl_cert]))
+        elsif options[:ssl_cert_string]
+          context.cert = OpenSSL::X509::Certificate.new(options[:ssl_cert_string])
+        elsif options[:ssl_cert_object]
+          context.cert = options[:ssl_cert_object]
+        end
+      end
+
+      def set_key(context, options)
+        passphrase = options[:ssl_key_pass_phrase]
+        if options[:ssl_key]
+          context.key = passphrase ? OpenSSL::PKey.read(File.open(options[:ssl_key]), passphrase) :
+            OpenSSL::PKey.read(File.open(options[:ssl_key]))
+        elsif options[:ssl_key_string]
+          context.key = passphrase ? OpenSSL::PKey.read(options[:ssl_key_string], passphrase) :
+            OpenSSL::PKey.read(options[:ssl_key_string])
+        elsif options[:ssl_key_object]
+          context.key = options[:ssl_key_object]
+        end
+      end
+
+      def set_cert_verification(context, options)
+        context.verify_mode = OpenSSL::SSL::VERIFY_PEER
+        cert_store = OpenSSL::X509::Store.new
+        if options[:ssl_ca_cert]
+          cert_store.add_cert(OpenSSL::X509::Certificate.new(File.open(options[:ssl_ca_cert])))
```

Comment:
Change this line to `cert_store.add_file(File.open(options[:ssl_ca_cert]))`. Adding a *cert* will only read the first cert, while adding a *file* will read all the certs in the file (which is definitely what we want, so we can get the intermediate cert).

Reply:
You will also need `cert_store.add_file(File.open(options[:ssl_cert]))` so it can add the intermediate cert to the chain.

Reply:
This makes sense, although makes me feel like our cert set up is weird. Either way, I'll add a thing to make this possible 😄

Reply:
@rebehe you and Mike may have talked about this, but I'm thinking about adding another option (something like :ssl_ca_cert_paths that will contain a list of paths to cert files this will read in. Any objections? I'll be something like:

```ruby
elsif options[:ssl_ca_cert_paths]
  raise TypeError("Option :ssl_ca_cert_paths should be an array of cert file paths") unless options[:ssl_ca_cert_paths].is_a? Array
  options[:ssl_ca_cert_paths].each { |path| cert_store.add_file(path) }
```

something like that?

Reply:
Is there a need for more than one ca_cert path? I believe we put all our trusted certs into one file.

Reply:
Sorry, looking at your example you do:
add_file(ca_file)
add_file(intermediate_cert_file)

I want to provide a way to add 1..N of those without changing the way the existing ones work

Reply:
Ah, gotcha :) in that case, I recommend renaming the option to not include the word `ca`

Reply:
got it - I need to add the ca_file and the cert_file. Will update! :)